### PR TITLE
DS-3209 AIP Import: Extend accepted handles for supports()

### PR DESCRIPTION
**Pull request by mjmarttila**
Adds an additional supported handle format to the supports() method of
VersionedHandleIdentifierProvider to make behavior more in-line with the
DS 5.4 code.

https://jira.duraspace.org/browse/DS-3209

### CODE REVIEW

**Comment by terrywbrady** [comment]
Note: this PR was developed/tested at Georgetown and it resolved an AIP ingest issue locally.

**Comment by tdonohue** [comment]
+1 is just restoring code from 5.4 which seems to have been accidentally stripped out

**Comment by pnbecker** [comment]
Since DSpace 6 we have to VersionedHandleIdentifierProvider. We should enhance both. While doing so I thought it would be good to also enhance the method retrieveHandleOutOfUrl. I put my changes here: https://github.com/tuub/DSpace/tree/DS-3209. Can you please review and test my addition and tell me what you think about that? If you like my additions, you could just cherry-pick them to the branch you used to create this PR, that would update this PR automatically.

**Comment by mjmarttila** [comment]
Thanks, @pnbecker. I looked through your suggested changes and thought they made sense to me. 

Unfortunately, with the way we use our environment, one of the lines would break the item import process for us. In our environment, we have one production instance and three non-prod instances of DSpace. Each separate instance has it's own handle prefix (e.g. 12345, 12345.1, 12345.2, 12345.3) that we use to help determine which items were created where. So the `getPrefix()` comparison on [line 353](https://github.com/tuub/DSpace/blob/DS-3209/dspace-api/src/main/java/org/dspace/identifier/VersionedHandleIdentifierProvider.java#L353) would then break any item imports from our other environments since those items would have slightly different handle prefixes.

I think we've been able to get away with our setup in previous versions of DSpace since the `retrieveHandleOutOfUrl()` method was slightly less stringent. I'm not sure if we're the only folks with that sort of a setup or if other institutions might be affected by that problem as well.

Does anybody else watching this PR have a similar setup or know of a similar setup? We may be a special case.

Has there ever been talk about the possibility of having a sort of regular expression handle.prefix configuration to provide more flexibility on this front?

**Comment by pnbecker** [comment]
Hi @mjmarttila 

I added the line in question to ensure, that we a) have a handle and b) that we have a handle we are responsible for. When I saw the message I became afraid that the VersionedHandleIdentifierProvider would accept URLs like http://dx.doi.org/10.14279/depositonce-5015 as handle. While DOI build up on handle, the VersionedHandleIdentifierProvider shouldn't deal with DOIs as if they were handles.

DSpace currently cannot deal with multiple different prefixes. I don't see how this could easily be achieved and to be honest, I don't see much sense in doing so. Of course I'm open for arguments... What I do understand is, that it would make sense to support the use case where someone is working with one prefix that is extended by a dot and something else. So instead of testing equality, we could use StringUtils.startsWithIgnoreCases(String, String). That would support the use-case of extending *one* prefix, but would still ensure we have a handle out of our range. Would that keep you compatible?

If my proposed solution fits your needs, can you extend your PR by cherry-picking my commit and adding another commit changing the equality checks or shall I create a new PR?

**Comment by mjmarttila** [comment]
Hi @pnbecker,

The reasons you’ve introduced the additional handle validation in that line certainly make sense, but unfortunately even with your suggested `startsWithIgnoreCases()` modification, the solution would break a multi-instance environment like ours.

It sounds like you already understand the value of using both a primary prefix and derived prefixes that extend it with a dot, but for the sake of the PR conversation history I’ll provide some additional context around why we use derived handle prefixes: when we import a production item into our test instances, we want to retain the production handle (e.g. XXX/YYY). When one of our test systems assigns a handle, we want to use a derived handle prefix (e.g. XXX.1/YYY) to indicate that the object originated on a particular test server. This way regardless of what content gets transferred where, we know which instance each object originated from.

So if we set the `handle.prefix` property on all of our test instances to our production handle with the `startsWithIgnoreCases()` addition in place, it would technically allow our test systems to import any of our existing exported AIPs. But it would also cause our test systems to then generate new handles using only the production prefix (e.g. XXX/YYY for all instances), breaking our system of determining where the content originated from.

If other institutions have a similar setup with derived handle prefixes, the solution as it stands would prevent those test handles from being assigned. What would folks think about something along the lines of a property in the local.cfg that the `supports()` method checks and, when set/not default, has the method split the configured handle prefix on the user-specified regex and only uses the first value for the prefix comparison? So in our case we could specify “[.]” as our expression which would then let the supports() check only look at the primary prefix but would not affect handle assignment during new object creation.

**Comment by pnbecker** [comment]
Hi @mjmarttila,

thanks for the context, that helps to understand what your aiming to. Are items that originated on your production server ever moved to the testservers? If not, your setup should even work with the startsWithIgnoreCase() comparison. In case you copy items originated on your production server to the test instances, could you please locally override this class using DSpaces "additions" module?

While thinking about how to solve this issue, I wrote the following code. But this requirement is so specific that I really hesitate to include this into DSpace's official code base.
```
public static String retrieveHandleOutOfUrl(String url)
    {
        // We can do nothing with this, return null
        if (!url.contains("/")) return null;

        String[] splitUrl = url.split("/");
        
        if (splitUrl.length >= 2)
        {
            // Handle prefixes can be extended by a "." and arbitrary numbers.
            // We need to ensuere here, that we found a handle containing a prefix
            // we are responsible for. Check this a little bit fuzzy to support
            // setups running multiple instances using an extended prefix.
            String retrievedPrefix = splitUrl[splitUrl.length -2];
            if (StringUtils.substringBefore(getPrefix(), ".").equalsIgnoreCase(
                    StringUtils.substringBefore(retrievedPrefix, ".")))
            {
                return retrievedPrefix + "/" + splitUrl[splitUrl.length - 1];
            }
        }
        return null;
    }
```

What does others think of this discussion?

**Comment by mjmarttila** [comment]
Hi @pnbecker,

Items originating from our production server do get transferred to our test servers so that we can test custom features that are in development on a slice of our production data before pushing those features to our prod instance.

If we're the only deployment using this sort of a multi-instance setup, then we could certainly do a local override (and thanks very much for providing a code example of how this might be done). We just thought we would use this opportunity to advocate for support of the multi-instance setup in case other institutions were working off of a similar model.

To that end, I'll echo your call for others to comment on this discussion.

**Comment by tdonohue** [comment]
To add to this discussion, there are known use cases for a production DSpace site actually having items with multiple handle prefixes (even though a single DSpace site only *assigns* a single handle prefix).  

One of the best examples is a *federated repository*, where one DSpace is actually harvesting / importing its content from other DSpace sites.  One example of this in production is the Texas Digital Library's (TDL) DSpace site: https://tdl-ir.tdl.org/tdl-ir/browse?type=title   If you look closely at the Item handles in that site, they come from many different Handle prefixes, as the original item exists in a *separate* DSpace.  In the case of TDL, their main repository just allows you to search the DSpace sites of all their member institutions.

It's also worth noting that many production DSpace sites assign handles that include dots in the prefix. So the fact that a prefix has a dot also shouldn't exclude it in any way.  Again, one example: http://oaktrust.library.tamu.edu/handle/1969.1/2815

So, long story short, I'm not sure we should be validating prefixes based on the configured `handle.prefix`.  That setting is only meant to represent what prefix DSpace will use when it *assigns* new handles. But, DSpace should still be able to import Items having existing handles (whether it be from AIP or OAI-PMH or wherever).

**Comment by pnbecker** [comment]
I read the protocol from yesterday's developer meeting (http://irclogs.duraspace.org/index.php?date=2016-05-25) and the comments in this ticket above. I hope I can clarify things a little bit:

- The PR by @mjmarttila is a good one and we should do the same change to the VerisonedHandleIdentifierProviderWithCanonicalHandles and to the HandleIdentifierProvider.
- While reviewing it, I became afraid that any HandleIdentifierProvider would return true on `supports("http://www.example.org/test123/arbitraryPath")` and return test123/arbitraryPath as handle. To say it simple: we shouldn't rely on dc.identifier.uri to contain handles only. ;-) As this PR connects the methods `support()` and `retriveOutOfURL()` we need to ensure that we found a handle.
- @mwoodiupui mentioned in the developer meeting that we need to ensure that the HandleProvider does not claim to support DOIs. That is true and we should keep that in mind. But I'm even more afraid of arbitrary URLs that could be stored in dc.identifier.uri and include two slashes or more.

Regarding a possible solution: I thought it would be a good idea to use the handle prefix to ensure that we have a handle there. But @mjmarttila and @tdonohue proofed that might have downsides. What can we use instead? Can we use handle.canonical.prefix or should we just check whether `/handle/` is part of the URL?

**Comment by mwoodiupui** [comment]
There are several cases to consider.

http[s]://hdl.handle.net/* contains a Handle.
hdl:* contains a Handle (but we should not be producing representations with this unregistered scheme).
http[s]://dspace.example.net/handle/* contains a Handle.
http[s]://www.example.net/dspace3/handle/* contains a Handle.
123.4/56 is a Handle.
10.14279/33 is a DOI (which is also a Handle, but IMHO we should not recognize it as one).

As far as I can tell, anything else is not a Handle and does not contain one.

Here's what I would do:
1)  If the identifier matches one of the "contains a Handle" rules, extract the part that should be a Handle.  Otherwise use the whole string.
2)  If what we have now starts with the DOI prefix, return false.
3)  If what we have now matches the rules for raw Handle representations, return true.
4)  Return false.

**Comment by mjmarttila** [comment]
Thanks to everybody for the clarification and thanks mwoodipui for the suggested solution outline. I've got the first draft of an implementation of that solution (still needs to be cleaned up):


```
// valid starting patterns for handle
    	String[] validPatterns = {
    			// http[s]://hdl.handle.net/*
    			"^http[s]?://hdl\\.handle\\.net/",
    			// hdl:*
    			"^hdl:",
    			// http[s]://dspace.example.net/handle/*
    			"^http[s]?://[a-zA-Z0-9/.\\-]+/handle/"
    		};
    	
    	// trim out prefix, if found
        String handle = identifier;
        boolean match = false;
        for(String vp : validPatterns) {
        	if(match)
        		break;
        	Pattern pattern = Pattern.compile(vp);
        	Matcher matcher = pattern.matcher(identifier);
        	if(matcher.matches()) {
        		handle = handle.substring(matcher.end());
        	}
        }
        
        // fail if identifier starts with DOI prefix
        if(handle.startsWith("10."))
        	return false;
        
        // fail if identifier does not match handle format
        if(! handle.matches("\\d+(\\.\\d+)*/\\d+(\\.\\d+)*"))
        	return false;
        else			
        	return true;
```

I do have a follow-up question before I clean up the code and add it to a commit, though.

@mwoodiupui, @pnbecker, @tdonohue, is there an established standard (from a DSpace perspective) of an acceptable handle value? The solution above assumes that all non-slash, non-dot characters will be numerical, but I also looked at the [Handle.net Tech Manual](http://www.handle.net/tech_manual/HN_Tech_Manual_8.pdf) (see Handle Syntax on page 6) which says, "Handles may consist of any printable characters from the Universal Character Set of ISO/IEC 10646, which is the exact character set defined by Unicode." Should DSpace be willing to accept any character from the UCS, or is there an agreed upon subset (such as the numerical assumption above).

Thanks for your help.

**Comment by pnbecker** [comment]
@mjmarttila Thanks for keep on working on this!

Regarding the patterns to detect handles, awful things like IDN (https://en.wikipedia.org/wiki/Internationalized_domain_name) came to my mind. => We cannot expect the URL to contain ASCII characters only. Perhaps it would be best to use something like http[s]?://.+/handle/ as third pattern?

We should support all characters from the UCS as the handle tech manual describes. DSpace currently produces numerical postfixes only, but the whole ticket just came up to support some kind of import workflow. Also it is possible that a handle contains more than one slash or perhaps once CNRI starts to use prefixes that contains characters... E.g. originally we didn't took care in DSpace about prefixes containing dots and had to resolve this later. Following this, we have to relax the check for the handle format to either "\\d+(\\.\\d+)*/.+" or even ".+/.+" (ensuring it contains at least a prefix, suffix and a slash).

In general I have the feeling that we are on a good way here! :-)

**Comment by tdonohue** [comment]
Quick, and perhaps silly question... Is all of this Handle validation parsing *really necessary*?  I'm not sure I understand why we couldn't just parse out the DOIs (so they are treated separately), and then treat everything else as "probably a Handle".  

I guess I'm not yet convinced all this fancy parsing is necessary. Is there something I'm missing that requires that we fully "validate" that a Handle is a Handle?  If so, why can't we simply check if it's a URL starting with "hdl:", "http://hdl.handle.net/" or whatever is configured in `handle.canonical.prefix`?

**Comment by pnbecker** [comment]
@mjmarttila: I discussed this with @tdonohue and @mwoodiupui in #dspace today. @tdonohue's last comment was related to the last five lines of your method: The form of a naked (=> without resolver or hdl: in front of it) handle is so arbitrary that it doesn't makes sense to check that. So we would suggest to relax the last one of the validPatterns (as described in my comment above), to add a pattern constructed by using handle.canonical.prefix from the configuration and to remove the last five lines. What do you think of those suggestions?

**Comment by mjmarttila** [comment]
Thanks @tdonohue, @pnbecker, and @mwoodiupui. I'm onboard with removing the last validation as unnecessary due to the fairly broad handle specification.

There is one more caveat that I found in the 11th hour and (briefly brought up on the #dspace IRC) by looking through the [stack trace of our original exception scenario](https://jira.duraspace.org/browse/DS-3209): in the case of an API import, the identifier string being supplied to the supports() method originates in the [getObjectHandle() method](https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/java/org/dspace/content/packager/AbstractMETSIngester.java#L1412) in the AbstractMETSIngester class, which essentially just pulls the OBJID from the METS file and strips out the "hdl:" prefix from the beginning. So the supports method should also be expecting the possibility of a valid handle without any of the other prefixes (e.g. "12345/6789").

I incorporated the suggestions outlined by @pnbecker and tried to reduce some of the regular expression complication to make the method more in-line with its original implementation's simplicity. The last four lines are a sort of high-level validation of handles coming in similarly to the API import stack I mentioned in the above paragraph. Basically, I just wanted to make sure it was meeting a very basic definition of a handle (contains a '/') and wasn't some other URL that did not match one of the known handle patterns (e.g. contained ".../handle/...").

```
        String prefix = handleService.getPrefix();
        String canonicalPrefix = DSpaceServicesFactory.getInstance().getConfigurationService().getProperty("handle.canonical.prefix");
        if (identifier == null)
        {
            return false;
        }
        // return true if handle has valid starting pattern
        if (identifier.startsWith(prefix)
                || identifier.startsWith(canonicalPrefix)
                || identifier.startsWith("hdl:")
                || identifier.startsWith("info:hdl")
                || identifier.matches("^https?://hdl.handle.net/")
                || identifier.matches("^https?://.+/handle/"))
        {
            return true;
        }
        // return false if identifier starts with DOI prefix
        if(handle.startsWith("10.") || handle.matches("^https?://dx.doi.org") || handle.startsWith("doi:"))
        	return false;
        // return false if identifier does not meet basic handle requirements (has a '/') 
        if(! handle.contains("/") || handle.matches("^https?://.+"))
        	return false;
        // otherwise, assumed a valid handle
                return true;
```

Would this meet your suggestions? Are the last four lines an acceptable way to handle the prefix-stripped handle input? Thanks as always for any feedback.

**Comment by mjmarttila** [comment]
I've added the code from my last comment into the latest commit. I'd be happy to incorporate any additional suggestions as they arise, though.

**Comment by terrywbrady** [comment]
@mjmarttila, during a PR review there was a request to provide detailed instructions for someone wishing to test this PR.  Please review/revise these instructions.

1. Export an AIP from an instance with handle prefix X
2. Restore that AIP into a DSpace 6 instance with a handle prefix of Y
3. Result: AIP should be restored and the handle from the original repo should be retained

**Comment by mjmarttila** [comment]
Thanks, @terrywbrady. I think your instructions look good to me. An example of the original exception that was thrown by the code in the scenario you described is available in the description of the accompanying JIRA ticket for this PR (see https://jira.duraspace.org/browse/DS-3209). With this PR in place, that exception should now only be thrown in the case of a non-recognizable identifier format (e.g. DOI format, non-handle URL, etc.).

If it helps, below is an explicit example of a test case:

1. An item AIP (with handle 12345.1/123456) is exported from SourceRepo with `handle.prefix:12345.1`.
2. That AIP is restored into DestRepo with `handle.prefix:12345.2`.
3. The item is successfully restored into DestRepo with its original handle value (12345.1/123456).

Note that since the change is being made to the import code of DS-6.0, I believe that the version of the DSpace instance that the AIP is being exported from shouldn't matter. I've been testing exporting from a v5.4 instance and restoring to a v6.0 instance.

If I can provide any additional clarification, please just let me know. Thanks in advance to any folks who test.

**Comment by terrywbrady** [comment]
@pnbecker , I see that you are assigned to this.  Would you have time to test this change?

**Comment by tdonohue** [comment]
+1 via code review. The code all looks reasonable. This is just waiting on a tester to re-verify (and @pnbecker volunteered to test in today's Dev meeting)

**Comment by terrywbrady** [comment]
+1, tested locally

**Comment by pnbecker** [comment]
Sorry, that I was so unresponsive in the last weeks.

I haven't tested it, but I'm pretty sure, that the code as it is now would return true for something like "http://dx.doi.org/10.14279/depositonce-5383" or even "doi:10.14279/depositonce-5383", which obviously aren't handles but DOIs. The problem is that we cannot assume that everything containing a slash and is not starting with "10." is a handle. And just think how this code would work together with other identifiers (not that we don't have enough, but that never prevented people from inventing new identifier systems).

I think we are putting the cart in front of the horse. Why do we try so hard to accept almost everything as handle? Instead we could change the AIP Import to **not** remove the information that a handle is a handle. If we keep the good code of this PR and change or even remove the method org.dspace.content.packager.AbstractMETSIngester.decodeHandleURN(String value) the handle recoginition and your use case should work fine.

The code would then look like following:
```
String prefix = handleService.getPrefix();
        String canonicalPrefix = DSpaceServicesFactory.getInstance().getConfigurationService().getProperty("handle.canonical.prefix");
        if (identifier != null &&
            (
                identifier.startsWith(prefix)
                || identifier.startsWith(canonicalPrefix)
                || identifier.startsWith("hdl:")
                || identifier.startsWith("info:hdl")
                || identifier.matches("^https?://hdl.handle.net/")
                || identifier.matches("^https?://.+/handle/")
            )
        )
        {
            return true;
        }
        return false;
```

Sorry for having objections and objections but this problem is bigger than it looks on the first wimp.

**Comment by tdonohue** [comment]
From talking with @pnbecker on IRC, it sounds like his primary objection (in the comment preceding this one) is that each of these methods return a [*default* value of `true`](https://github.com/DSpace/DSpace/pull/1399/files#diff-500149cde8c51bb626c8d5dc45c4d594R76).  Returning "true" by default essentially means that it's going to match a lot of URLs that are *not* handles (or DOIs)... e.g. "http://example.com/" would return true.

Simply setting these methods to a default value of `false` would both simplify the method (as @pnbecker shows above) and ensure we are not overly "greedy" with our matching of what is a valid handle.

The sidenote by @pnbecker about changing the behavior of AIP import may unravel a larger problem here. The reason the "hdl:" seems to be stripped from the handle is that the handle string ends up being passed to [PackageUtils.createObject()](https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/java/org/dspace/content/packager/PackageUtils.java#L460), which in turn creates the necessary object (e.g. via [CollectionService.create()](https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/java/org/dspace/content/CollectionServiceImpl.java#L83)).  Based on the *type* of object, it may or may not use the IdentifierService to assign the Handle. Currently, it seems only *Items* use IdentifierService in [ItemInstallService.installItem()](https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/java/org/dspace/content/InstallItemServiceImpl.java#L60), whereas Communities and Collections simply call the [HandleService.createHandle()](https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/java/org/dspace/handle/HandleServiceImpl.java#L161) which calls the HandleDAO to insert it directly into the database tables.

So, that previous paragraph is a long way of saying, I think the IdentifierService is *not* yet fully integrated into every layer of our API.  There are areas that simply insert Handles directly using HandleService (& therefore HandleDAO), especially Collections and Communities, instead of having the "matching" IdentifierService 'mint' a new Handle.  Since AIPs have to work for *all* types of objects, we end up having to "work around" this problem until IdentifierService fully replaces HandleService.

@mjmarttila and @terrywbrady : do you see any issue with simply changing these methods to "return false;" as the default (instead of true)?  Just curious if that's a reasonable "compromise" for now.

**Comment by mjmarttila** [comment]
@pnbecker and @tdonohue, I believe that all of the examples you have cited ("http://dx.doi.org/10.14279/depositonce-5383", "doi:10.14279/depositonce-5383", and "http://example.com") appropriately return false with the logic as it stands (see [Line 71](https://github.com/Georgetown-University-Libraries/DSpace/blob/12649fe1fe9fa73342fa424183cbb54ce12049bf/dspace-api/src/main/java/org/dspace/identifier/HandleIdentifierProvider.java#L71) for examples one and two, and [Line 74](https://github.com/Georgetown-University-Libraries/DSpace/blob/12649fe1fe9fa73342fa424183cbb54ce12049bf/dspace-api/src/main/java/org/dspace/identifier/HandleIdentifierProvider.java#L74) for example three). (Although I see now that the pattern for DOI site matches on Line 71 was a bit off and should be corrected to "^https?://dx\\.doi\\.org.*" .) The method takes the approach of returning false for known non-DSpace namespaces (the various DOI prefixes, arbitrary URLs, etc.) and then returning true for anything that passes those tests and still lies within the rather large handle namespace. This was meant to accommodate handle.net's very loose definition of a handle in the technical documentation while still avoiding the known non-DSpace namespaces.

That being said, I'm fine going with the more restrictive approach recommended, but would suggest a small addition to allow for multi-instance DSpace deployments that utilize derived handles. The method plus addition would look like the following:
```
        String prefix = handleService.getPrefix();
        String canonicalPrefix = DSpaceServicesFactory.getInstance().getConfigurationService().getProperty("handle.canonical.prefix");
        if (identifier == null)
        {
            return false;
        }
        // return true if handle has valid starting pattern
        if (identifier.startsWith(prefix)
                || identifier.startsWith(canonicalPrefix)
                || identifier.startsWith("hdl:")
                || identifier.startsWith("info:hdl")
                || identifier.matches("^https?://hdl\\.handle\\.net/")
                || identifier.matches("^https?://.+/handle/"))
        {
            return true;
        }
        // return true if base prefix matches in case of multi-instance deployment with derived prefixes demarcated by a dot "." 
        if(prefix.contains(".")) {
        	String[] splitPrefix = prefix.split("\\.");
        	if(splitPrefix.length > 1 && identifier.startsWith(splitPrefix[0])) {
        		return true;
        	}
        }
        return false;
```

Would this satisfy the objections? If so, I'll add a new commit with the changes.

I agree that the root issue is larger than just the supports() method, but created this PR and accompanying ticket just to try to bring the handle behavior more in-line with its behavior in v5.4. I'm afraid I'm not familiar enough with the inner mechanics of all the affected classes to go making the larger changes that will probably be required to get the object ingest/creation processes where they eventually should be.

**Comment by mjmarttila** [comment]
I've incorporated the suggestions for a more restrictive supports() method into the [latest commit](https://github.com/DSpace/DSpace/pull/1399/commits/f6bfc491d391bdba62b5a96c4e33c1b1abb4dd18).

**Comment by tdonohue** [comment]
@pnbecker : Could you take a look at these latest changes and give your feedback?  To me, they sound more in line with the solution you were looking for.  It'd obviously be good to get this ticket resolved (as it's been a long standing one)

**Comment by mwoodiupui** [comment]
We have continuing uncertainty about the set of recognized forms because we have no unit test suite to prove our expectations.  I would suggest putting the list of test cases in an external file so that it can be easily adjusted without rebuilding.

**Comment by mwoodiupui** [comment]
See https://github.com/mwoodiupui/DSpace/tree/DS-3209 for a sketch of a test suite.  Unfortunately I don't seem to be able to create a PR against the Georgetown repo, or I set mine up wrong.  The .properties resource can be edited as desired to set the single test's expectations.  The test can be run singly:  "mvn test -Dtest=HandleIdentifierProviderTest" for rapid turnaround.

That patch also makes small changes to the ServiceManager so that the test can bypass the Spring configuration and work with the specific flavor of provider that it expects, still autowired by Spring.  In this way we could test all flavors without any reconfiguration of Spring.  I didn't like making unrelated changes deep in dspace-services just for testing, but I couldn't work out a better approach.  The updated code should be equivalent to the original, except that the resulting ApplicationContext can receive new bean definitions later.  If this change is objectionable, we can just use it for testing identifier patterns during development and leave it out of the final merge.

**Comment by terrywbrady** [comment]
@mwoodiupui , I merged your branch into @mjmarttila 's branch. Thank you for providing this.

Mike is out of the office for a few days.  I will plan to test this locally to confirm functionality.

**Comment by mwoodiupui** [comment]
Ha, shortly after posting my patch, I thought of a less intrusive way to do it.  The branch mentioned above has a new commit which reverses the changes to SpringServiceManager, replacing them with a new child ApplicationContext confined to the test suite.

Ugh, I just noticed that the .properties needs to have colons (such as "http:") escaped.  As I said, a sketch.

File: dspace-api/src/test/resources/org/dspace/identifier/handle-forms.properties [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Key is the identifier to be tested; value is "true" or "false":  should this
+# identifier be supported by the provider under test?
+
+http://dx.doi.org/10.14279/depositonce-5383 = false
+123456789/1 = true
```
Comment (tdonohue) [inline dspace-api/src/test/resources/org/dspace/identifier/handle-forms.properties]: Seems like a good direction to add in these Unit Tests to prove out that we are handling all identifiers appropriately.  We may want to add a note that `123456789/1 = true` ONLY succeeds in this scenario because the [Unit Test framework's `handle.prefix=123456789` in its local.cfg](https://github.com/DSpace/DSpace/blob/master/dspace-api/src/test/data/dspaceFolder/config/local.cfg#L68).

To that end, we should enhance this list with other examples that should/should not be supported.  Here's a few that come to mind:

```
# Note: This next one only succeeds because handle.prefix=123456789 in Test Framework's local.cfg
123456789.1/1 = true

# Other matches that should succeed
hdl:123456789/1 = true
hdl:123.456/1.2 = true
info:hdl/12345/4000 = true
http://hdl.handle.net/20.500.1/1.2 = true
https://hdl.handle.net/987/1 = true
http://my.dspace.edu/handle/123/456 = true
https://my.dspace.edu/handle/1.23/45.6 = true

# Other matches that should fail
http://example.com/ = false
987654321/1 = false
12.3456789/1 = false
```

File: dspace-api/src/main/java/org/dspace/identifier/HandleIdentifierProvider.java [hunk 2]
```diff
@@ -50,21 +51,30 @@ public boolean supports(Class<? extends Identifier> identifier) {
...
             return true;
         }
-        
+        // return true if base prefix matches in case of multi-instance deployment with derived prefixes demarcated by a dot "." 
+        if(prefix.contains(".")) {
+            String[] splitPrefix = prefix.split("\\.");
+            if(splitPrefix.length > 1 && identifier.startsWith(splitPrefix[0])) {
+                return true;
+            }
+        }
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/identifier/HandleIdentifierProvider.java]: I just realized this logic may be slightly flawed for brand new Handle prefixes.  For those who haven't heard, when you buy a new prefix from the Handle.net Registry, they are now of the format `20.500.xxxxx` (i.e ALL handle prefixes now begin with "20.500.*").  This means that newer "derived" prefixes will end up being of the format `20.500.xxxx.yyy`, see this FAQ question on their site for an example: https://www.handle.net/faq.html#4

Unfortunately, the logic above may not work right for new prefixes, as it seems to assume there will only be *one* period within the prefix.  To be *more correct* this probably should be matching against the *last period* in the prefix.. something like

```
       if(prefix.contains(".")) {
            String prefixBeforeLastPeriod = prefix.substring(0, prefix.lastIndexOf(".")+1);
            if(prefixBeforeLastPeriod.length > 1 && identifier.startsWith(prefixBeforeLastPeriod)) {
                return true;
            }
        }
```
This still isn't "perfect" , as a configuring a prefix of "20.500.1000" will also cause a prefix of "20.500.88" to return "true".  But, I'm not sure there's any way to make this 100% perfect.

File: dspace-api/src/test/resources/org/dspace/identifier/handle-forms.properties (hunk 1 above)
Comment (mwoodiupui) [inline dspace-api/src/test/resources/org/dspace/identifier/handle-forms.properties]: I had been thinking about "...only succeeds because...."  Actually, a proper unit test would set handle.prefix and handle.canonical.prefix to known values itself, before calling the target code.

Comment (tdonohue) [inline dspace-api/src/test/resources/org/dspace/identifier/handle-forms.properties]: @mwoodiupui : I agree. Which is why I wanted to point out *why* that test is successful.  Either we need to build this into a better test, or we should minimally add a comment to that `handle-form.properties` file to explain why that test is successful (in case someone should purposefully update the `local.cfg` for the Test Framework only to find this test failing)

**Comment by terrywbrady** [comment]
Discussed this in IRC with @tdonohue and he pointed out the following config setting: https://github.com/DSpace/DSpace/blob/master/dspace/config/dspace.cfg#L255

I will attempt to use this setting rather than coding for a derived prefix.

**Comment by terrywbrady** [comment]
I made an update per @tdonohue 's suggestion to use the handle.additional.prefixes config setting.

I have tested this locally and the settings seem to validate properly.

**Comment by tdonohue** [comment]
@terrywbrady : looking good. I notice we still should update the Unit Test scenarios in `handle-form.properties` (noted in this comment: https://github.com/DSpace/DSpace/pull/1399#discussion_r74146331).  That'd just help us prove this is working properly

**Comment by terrywbrady** [comment]
Latest commit contained only test driver changes

**Comment by terrywbrady** [comment]
@tdonohue , @mjmarttila , @mwoodiupui , @pnbecker , test cases have been updated and ran successfully in travis.

**Comment by tdonohue** [comment]
+1 via code review.  Looks good to me, passes all the new unit test scenarios.

**Comment by pnbecker** [comment]
This looks really great now. Using the property handle.additional.prefixes seems to be a perfect solution for me. Thanks for all your hard work on this!
+1 => merging.
